Network/Socket/Types.hsc

```diff
@@ -920,38 +919,28 @@ withSockAddr addr f = do
 -- | Write the given 'SockAddr' to the given memory location.
 pokeSockAddr :: Ptr a -> SockAddr -> IO ()
 #if defined(DOMAIN_SOCKET_SUPPORT)
-pokeSockAddr p (SockAddrUnix path) = do
-# if defined(darwin_HOST_OS)
-    zeroMemory p (#const sizeof(struct sockaddr_un))
-# else
-    case path of
-      ('\0':_) -> zeroMemory p (#const sizeof(struct sockaddr_un))
-      _        -> return ()
-# endif
+pokeSockAddr p sa@(SockAddrUnix path) = do
+    zeroMemory p $ fromIntegral $ sizeOfSockAddr sa
 # if defined(HAVE_STRUCT_SOCKADDR_SA_LEN)
     (#poke struct sockaddr_un, sun_len) p ((#const sizeof(struct sockaddr_un)) :: Word8)
 # endif
     (#poke struct sockaddr_un, sun_family) p ((#const AF_UNIX) :: CSaFamily)
     let pathC = map castCharToCChar path
-        poker = case path of ('\0':_) -> pokeArray; _ -> pokeArray0 0
-    poker ((#ptr struct sockaddr_un, sun_path) p) pathC
+    -- the buffer is already filled with nulls.
+    pokeArray ((#ptr struct sockaddr_un, sun_path) p) pathC
```

Comment:
This might not be safe when `path` is unexpectedly long.  We now have `sizeOfSockAddr` for unix-domain sockets as a constant `sizeof(sockaddr_un)`, and it seems possible that trying to use a longer path will corrupt memory?

Reply:
I tested this with a short program that calls `allocaBytes 128` and then `pokeArray p` with an array of `> 128` C chars.  The compiler and run-time did not object, and dutifully dumped core when the array was large enough.  So here, we need to fail if `SockAddrUnix path` contains a `path` that is too long (which is a bit shorter than 128 bytes, since we have to account for the `sun_family` structure member. It may be unwise to assume that it is the standard 2 bytes, so we could either compute the overhead explicitly, or just cap the size at `120` (again as a named constant, properly commented at the top of the file).

Reply:
Now, throw an error.